Add rustls_str, rustls_slice_bytes, and friends

jsha opened the pull request:
For the basic slice types we need, this introduces `rustls_str`, `rustls_slice_bytes`, and `rustls_slice_u16`. For slices of slices, we have `rustls_slice_str` and `rustls_slice_slice_bytes`. Each of these types have a lifetime parameter that ensures they don't outlive the object they are borrowing from. 

The `str` types have an additional guarantee on top of Rust's guarantee that strings contain UTF-8: They can't contain `\0`. This makes it possible to interpolate them into C strings and compare them against C strings using `strncmp`.

Built off @icing's #50.

tgeoghegan commented on src/rslice.rs:
```diff
@@ -0,0 +1,184 @@
+}
+
+impl<'a> From<&'a [u8]> for rustls_slice_bytes<'a> {
+    fn from(s: &[u8]) -> Self {
```
My intuition is that you need `s` (which is a reference) to at least outlive `'a`. Is that lifetime bound implicit here?

jsha replied:
Yep, this is lifetime elision. https://doc.rust-lang.org/nomicon/lifetime-elision.html

> If there is exactly one input lifetime position (elided or not), that lifetime is assigned to all elided output lifetimes.

divergentdave replied:
Moreover, the lifetimes are defined to be equal in the impl line above (the lifetime of the slice that's the parameter to From, and the lifetime parameter of rustls_slice_bytes), and the compiler will check that the function signature here (after applying lifetime elision rules) matches the signature of From::from with the type parameters substituted in.

tgeoghegan commented on src/rslice.rs:
```diff
@@ -0,0 +1,184 @@
+/// directly because `Vec` is not #[repr(C)]. So we build a new Vec containing
+/// rustls_slice_bytes<'a> as an owned object. We can then expose slices of
+/// that Vec to C.
+pub(crate) struct VecSliceBytes<'a>(Vec<rustls_slice_bytes<'a>>);
```
Could this be a type alias `type VecSliceBytes = Vec<rustls_slice_bytes<'a>>;` and save us the enclosing envelope struct?

jsha replied:
Unfortunately you can't implement traits on type aliases. That's one of the main reasons for using the newtype pattern: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types

tgeoghegan commented on src/rslice.rs:
```diff
@@ -0,0 +1,184 @@
+pub struct rustls_slice_slice_bytes<'a> {
+    data: *const rustls_slice_bytes<'a>,
+    len: size_t,
+    phantom: PhantomData<&'a [&'a [u8]]>,
```
nit: could be `PhantomData<&'a [rustls_slice_bytes<'a>]>` for consistency, but IIUC it doesn't matter what goes in the angle brackets as long as it includes `'a`.

tgeoghegan commented on src/rslice.rs:
```diff
@@ -0,0 +1,184 @@
+
+/// An owned Vec<rustls_str>, where the inner slices have lifetime 'a.
+/// See comment on VecSliceBytes for more information.
+pub(crate) struct VecStr<'a>(Vec<rustls_str<'a>>);
```
It seems like `VecStr` and `VecSliceBytes` are manual monomorphizations of `Vec<T>`. But I think `VecStr` (for instance) only gets used in `rustls_slice_str::from<&'a VecStr<'a>>`, so it's never exposed to C. What I wonder is: could you implement `From<&'a Vec<rustls_str<'a>>` on `rustls_slice_str`, move the implementation of `VecStr::new` into `rustls_slice_str::from` and eliminate the need for `VecStr` altogether? I think that's pretty much what you did for `rustls_slice_u16`.

jsha commented:
> Could VecStr and VecSliceBytes be eliminated?

You also mention this in https://github.com/abetterinternet/crustls/pull/54#discussion_r585039954, so I'll respond here: Unfortunately they can't. We would get a "temporary value dropped here while still borrowed." That is, we need to create a new owned `Vec` that holds `rustls_str`'s instead of `&str`s. Given we need an owned object, who owns it? If we create it within a `From` implementation, it will be dropped before `from` returns, so `rustls_slice_str` can't hold a reference to it. Ultimately, whatever function is going to pass this off into C needs to hold the owned `Vec`, to ensure that it lives long enough for the callback to complete.

> Why don't rustls_str etc. appear in an updated crustls.h?

[cbindgen publishes functions, globals, and constants](https://github.com/eqrion/cbindgen/blob/master/docs.md#writing-your-c-api) - plus any types required by those functions, globals and constants. Since these types aren't (yet) used by any published functions, cbindgen does not publish them. You can see rendered .h files [here](https://github.com/abetterinternet/crustls/pull/50/files#diff-d78a232a38a96236f8e8bbe0acb353b20997674d6555e2ab75b16cbc0348d9ecR139-R214) and [here](https://github.com/abetterinternet/crustls/pull/51/files#diff-d78a232a38a96236f8e8bbe0acb353b20997674d6555e2ab75b16cbc0348d9ecR144-R196).

tgeoghegan commented:
Thanks for the clarifications and links; re-approved!

alex commented on src/rslice.rs:
```diff
@@ -0,0 +1,184 @@
+/// that Vec to C.
+pub(crate) struct VecSliceBytes<'a>(Vec<rustls_slice_bytes<'a>>);
+
+impl<'a> VecSliceBytes<'a> {
```
I'm a bit confused on how this is intended to be used, can you sketch out an example?

I also don't love needing to allocate for this. It seems preferable to keep the original `Vec<Vec<u8>>` and have C functions to extract a `rustls_slice_bytes` at a given index?

jsha replied:
One example here: https://github.com/abetterinternet/crustls/pull/51/files#diff-7f93c4e263c4e9ec748f804c7fd04a3b2fde86ffd741fb5516d67e1097bae4c1R149-R183

And one here: https://github.com/abetterinternet/crustls/pull/50/files#diff-48d27698cc708c7efe770fd897e4885efaa0a15cae30d54936068603ba03bbd9R696-R711

Good point on accessor functions! I recall that you had recommended that before and I forgot to pursue it. I'll give that a try.

jsha replied:
For the accessor version, I'm thinking of a function signature like this:

```
pub extern "C" fn rustls_slice_slice_bytes_get(
    input: rustls_slice_slice_bytes, n: usize) -> rustls_slice_bytes {
```

But this requires there be some sort of sentinel value of `rustls_slice_bytes` that means "n was too big." I guess that can be `rustls_slice_bytes{data: null(), len: 0}`. Interested in other ideas! I could also do:

```
pub extern "C" fn rustls_slice_slice_bytes_get(
    input: rustls_slice_slice_bytes, n: usize, out: *mut rustls_slice_bytes) -> rustls_result {
```

But that feels pretty awkward to use in a loop.

divergentdave replied:
I agree on the accessor recommendation. They would allow encapsulating some memory layouts, cutting down on the number of types needed on the Rust size, plus calls to the accessors could be inlined when using a cross-language LTO-capable toolchain, making this a cheap abstraction.

The first accessor signature looks good to me, NULL checks seem like idiomatic C IMHO.

divergentdave commented on src/rslice.rs:
```diff
@@ -0,0 +1,188 @@
+    fn from(s: &[u16]) -> Self {
+        rustls_slice_u16 {
+            data: s.as_ptr() as *const u16,
+            len: s.len() as size_t,
```
FYI, both these casts are redundant, and the from/new functions aren't consistent on whether they do or don't cast.

jsha commented:
I've pushed an update with the accessor style. It's not correct yet - see the segfault in https://github.com/abetterinternet/crustls/pull/51 when I tried to incorporate it into that branch. But I figured I'd push up what I've got so far for folks to take a look!

I suspect I'm mixing up types between a thin pointer (`*const T`) and a fat pointer (`&[T]`), because the pointer value is 1, which happens to also be the length of the particular chain we're dealing with.

```
Program received signal SIGSEGV, Segmentation fault.
0x0000555555580426 in crustls::rslice::rustls_slice_slice_bytes_get (input=0x7fffffffa848, n=0) at /home/jsha/crustls/src/rslice.rs:95
95              Some(rsb) => (*rsb).into(),
(gdb) bt
#0  0x0000555555580426 in crustls::rslice::rustls_slice_slice_bytes_get (input=0x7fffffffa848, n=0) at /home/jsha/crustls/src/rslice.rs:95
#1  0x00005555555800d2 in verify (userdata=0x55555573b425, params=0x7fffffffaa48) at src/main.c:487
(gdb) p rsb
$1 = (*mut &[u8]) 0x1
```

divergentdave commented:
The first thing I noticed is that `rustls_slice_slice_bytes` is now a ZST, since it only contains a PhantomData. I think the issue is that `rustls_slice_slice_bytes_new` is returning a stack pointer to the reference it was passed in. It first takes a reference to the input slice reference, then casts it to one pointer type, then another. Passing an opaque pointer to either a slice (a fat pointer itself) or a Vec does sound like a promising way to go, but we'll need to be careful that the pointee stays put while it is in use.

jsha commented:
Good catch on the stack pointer, thanks! Pushed a fix. I thought of another approach: we could make `rustls_slice_slice_bytes` a newtype holding `&[&[u8]]`, but make it opaque to C by not making it `pub` (which causes cbindgen to generate an opaque struct declaration).

jsha commented:
Alright, I switched to a newtype pattern for `_slice_slice_bytes` and `_slice_str`, and I like it a lot better. I don't have to do nearly as much casting of thing to `*const T`. I've also added some tests, and updated #51 to use the latest version of `rslice`. Ready for another round of reviews!

icing commented on src/crustls.h:
```diff
@@ -268,6 +340,33 @@ void rustls_error(enum rustls_result result, char *buf, size_t len, size_t *out_
+ * Retrieve the nth element from the input slice of slices. If the input
+ * pointer is NULL, returns 0.
+ */
+size_t rustls_slice_slice_bytes_len(const struct rustls_slice_slice_bytes *input);
```
Comment is copy+pasta. Been there, done that.

tgeoghegan replied:
This is odd, because the corresponding doccomment in `rslice.rs` is correct.

icing commented on src/crustls.h:
```diff
@@ -268,6 +340,33 @@ void rustls_error(enum rustls_result result, char *buf, size_t len, size_t *out_
+ * Retrieve the nth element from the input slice of slices. If the input
+ * pointer is NULL, returns 0.
+ */
+size_t rustls_slice_str_len(const struct rustls_slice_str *input);
```
Description.

icing reviewed (approved):
Nice design the never construct the vec of struct thing, but to pass out views on the elements as value!

divergentdave reviewed (commented):
FYI, I did a Miri spot-check of the tests here, and they passed. I changed the crate-type in Cargo.toml from staticlib to lib, then ran `cargo +nightly miri test`.

jsha commented:
> FYI, I did a Miri spot-check of the tests here, and they passed. I changed the crate-type in Cargo.toml from staticlib to lib, then ran cargo +nightly miri test.

Why change from staticlib to lib? When I run with `crate-type = ["staticlib"]` things also work.

divergentdave commented:
Ah, when I did that at first, it didn't pick up any test targets. Perhaps it was fixed recently, and my nightly channel as installed was too old to have the fix.

tgeoghegan commented on src/rslice.rs:
```diff
@@ -0,0 +1,305 @@
+                return rustls_slice_bytes {
+                    data: null(),
+                    len: 0,
+                    phantom: PhantomData,
```
nit: you could save yourself a little bit of repetition by providing an `std::Default` implementation that constructs a "zero valued" `rustls_slice_bytes`.

jsha replied:
This is a good point, but I've found on another project that `std::Default` hides things in a way that can cause trouble. I'd like to favor explicitness and obviousness on this project.

tgeoghegan reviewed (approved):
Couple nits and a doccomment remark but LGTM!

jsha commented:
on `crustls.h`:
> This is odd, because the corresponding doccomment in rslice.rs is correct.

I hadn't updated `crustls.h`, and also the test phase that was supposed to check for that was broken (I needed `--exit-code` on `git diff`)

icing commented:
\o/
